robot/impl/local_robot_test.go

```diff
@@ -4284,3 +4284,147 @@ func TestMaintenanceConfig(t *testing.T) {
+
+	// Set up a local main robot which is connected to the remote.
+	mainRobotCfg := &config.Config{
+		Remotes: []config.Remote{
+			{
+				Name:    "remote",
+				Address: remoteAddr,
+				// These values dictate how quickly we'll observe the remote going offline. And how
+				// quickly we'll observe it coming back online.
+				ConnectionCheckInterval: 10 * time.Millisecond,
+				ReconnectInterval:       10 * time.Millisecond,
+			},
+		},
+	}
+	mainRobotI := setupLocalRobot(t, ctx, mainRobotCfg, logger.Sublogger("main"))
+	// We'll manually access the resource manager to move the test forward.
+	mainRobot := mainRobotI.(*localRobot)
+
+	// Grab the RobotClient resource graph node from the main robot that is connected to the
+	// remote. We'll use this to know when the main robot observes the remote has gone offline.
+	mainToRemoteClientRes, _ := mainRobot.RemoteByName("remote")
+	test.That(t, mainToRemoteClientRes, test.ShouldNotBeNil)
+	mainToRemoteClient := mainToRemoteClientRes.(*client.RobotClient)
+	test.That(t, mainToRemoteClient.Connected(), test.ShouldBeTrue)
+
+	// Stop the remote's web server. Wait for the main robot to observe there's a connection problem.
+	logger.Info("Stopping web")
+	remoteRobot.StopWeb()
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		tb.Helper()
+		test.That(tb, mainToRemoteClient.Connected(), test.ShouldBeFalse)
+	})
+
+	// Reconfigure the main robot with the offline remote removed
+	mainRobot.Reconfigure(ctx, &config.Config{})
+
+	// Ensure that the remote has been remoted correctly
+	findRemote, ok := mainRobot.RemoteByName("remote")
+	test.That(t, findRemote, test.ShouldBeEmpty)
+	test.That(t, ok, test.ShouldBeFalse)
+
+	// Ensure that motor is removed
+	names := mainRobot.ResourceNames()
+	test.That(t, names, test.ShouldNotContain, motorResourceName)
+}
+
+// TestRemovingOfflineRemotes tests a case where a robot's reconfigure loop is modifying
+// a resource graph node at the same time as the complete config loop. In this case the remote is
+// marked to be removed by the reconfig loop and then marked unhealthy by the complete config
+// loop. This caused the remote that should have been removed to be stay on the robot
+// and continue to try and reconnect. We recreate that scenario and ensure that our fix
+// prevents that behavior and removes the remote correctly.
+func TestRemovingOfflineRemotes(t *testing.T) {
+	// Close the robot to stop the background workers from processing any messages to triggerConfig
+	r := setupLocalRobot(t, context.Background(), &config.Config{}, logging.NewTestLogger(t))
+	r.Close(context.Background())
+	localRobot := r.(*localRobot)
+
+	// Create a context that we can cancel to similuate the remote connection timeout
+	ctxCompleteConfig, cancelCompleteConfig := context.WithCancel(context.Background())
+
+	// This cancel is used to ensure the goroutine is cleaned up properly after the test
+	ctxReconfig, cancelReconfig := context.WithCancel(context.Background())
```

Comment:
add defer cancels here so that if anything errors out or fails an assertion before the end of the test, we still end up being able to cleanup